src/Aspire.Hosting/ApplicationModel/ResourceLoggerService.cs

```diff
@@ -220,45 +238,46 @@ public event Action<bool> OnSubscribersChanged
         /// <returns>The log stream for the resource.</returns>
         public async IAsyncEnumerable<IReadOnlyList<LogLine>> WatchAsync([EnumeratorCancellation] CancellationToken cancellationToken = default)
         {
-            var channel = Channel.CreateUnbounded<LogLine>();
+            // Line number always restarts from 1 when watching logs.
+            // Note that this will need to be improved if the log source (DCP) is changed to return a maximum number of lines.
+            var lineNumber = 1;
+            var channel = Channel.CreateUnbounded<InternalLogLine>();
 
             using var _ = _logStreamCts.Token.Register(() => channel.Writer.TryComplete());
 
-            LogLine[]? backlogSnapshot = default;
-            var flushBacklogSync = new ManualResetEventSlim();
-            void Log(LogLine log)
+            InternalLogLine[]? backlogSnapshot = null;
+            void Log(InternalLogLine log)
             {
-                if (flushBacklogSync.IsSet)
-                {
-                    channel.Writer.TryWrite(log);
-                    return;
-                }
-
-                flushBacklogSync.Wait(cancellationToken);
-                // We need to ensure we don't write this log to the channel if it was already in the backlog
-                if (backlogSnapshot?.Contains(log) == false)
+                lock (_backlog)
                 {
-                    channel.Writer.TryWrite(log);
+                    // Don't write to the channel until the backlog snapshot is accessed.
+                    // This prevents duplicate logs in result.
+                    if (backlogSnapshot != null)
+                    {
+                        channel.Writer.TryWrite(log);
+                    }
                 }
             }
-
-            // From the moment we add this callback, logs will be written to the backlog & to our Log method above
-            // so our Log method needs to ensure it de-dupes logs.
             OnNewLog += Log;
 
-            backlogSnapshot = GetBacklogSnapshot();
-            flushBacklogSync.Set();
+            // Add a small delay to ensure the backlog is replayed from DCP and ordered correctly.
+            await EnsureBacklogReplayAsync(cancellationToken).ConfigureAwait(false);
+
+            lock (_backlog)
+            {
+                backlogSnapshot = GetBacklogSnapshot();
```

Comment:
`GetBacklogSnapshot()` already locks inside of it. There isn't a reason to lock here (in an async method) as well.